### PR TITLE
Jh/parser

### PR DESCRIPTION
**Pull request by aethos**
update the parser

### CODE REVIEW

File: src/parser.mly [hunk 1]
```diff
@@ -1,27 +1,245 @@
...
+  | fdecl_list fdecl { List.rev $1 }
+
+/* TODO: right now functions need to have variables declared before statements */
+fdecl:
+  TYPE_DEF ID lambda { $2, $3 }
+  | TYPE_DEF ID LPAREN formals_opt RPAREN FUNC_RET_TYPE typ
+    ASSIGN LBRACE stmt_list RBRACE
+      { { func_name = $2; function_return_t = $7;
+      function_formals = $4; function_body = $10;
+      function_return = $} }
```
Comment (vlad-scherbich) [inline src/parser.mly:82]: `function_return` seems to be missing a var. Let's fix this before merging.

**Review by vlad-scherbich — APPROVED** [review]

File: src/parser.mly [hunk 2]
```diff
@@ -1,27 +1,245 @@
...
+  | expr ARITH_MOD    expr                          { bin_op($1, Mod, $3) }
+  | expr LOGIC_EQ     expr                          { bin_op($1, Equal, $3) }
+  | expr LOGIC_NEQ    expr                          { bin_op($1, Neq, $3) }
+  | expr LCARET       expr                          { bin_op($1, Less, $3) }
+  | expr LOGIC_LEQ    expr                          { bin_op($1, Leq, $3) }
+  | expr RCARET       expr                          { bin_op($1, Greater,$3) }
+  | expr LOGIC_GEQ    expr                          { bin_op($1, Geq, $3) }
+  | expr LOGIC_AND    expr                          { bin_op($1, And, $3) }
+  | expr LOGIC_OR     expr                          { bin_op($1, Or, $3) }
+  /* | MINUS expr %prec NEG                         { Unop(Neg, $2) } */
```
Comment (vlad-scherbich) [inline src/parser.mly:230]: Should this be uncommented?

File: src/parser.mly [hunk 3]
```diff
@@ -1,27 +1,245 @@
...
+  /* | NOT expr                                     { Unop(Not, $2) } */
+  | ID LPAREN actuals_opt RPAREN                    { Call($1, $3) }
+  | LPAREN expr RPAREN                              { $2 }
+  | ID LPAREN actuals_opt RPAREN ACT_SEND ID        { None }
+  | ID LPAREN actuals_opt RPAREN ACT_BROADCAST ID   { None }
+  /* TODO: implement ACT_BROADCASTing lists of messages */
+  /* NOTE: negation, eg !a, does not exist in our LRM */
```
Comment (vlad-scherbich) [inline src/parser.mly:237]: Is this by design?

File: src/parser.mly [hunk 4]
```diff
@@ -1,27 +1,245 @@
...
-    | VARIABLE ASN expr { Asn($1, $3) }
-    | VARIABLE          { Var($1) } 
-    | LITERAL           { Lit($1) }
+  ID                                                { Id($1) }
+  | LITERAL                                         { Lit($1) }
+  | DOUBLE_LIT                                      { Double_Lit($1) }
+  | CHAR_LIT                                        { Char_Lit($1) }
+  | STRING_LIT                                      { String_Lit($1) }
+  | BOOL_LIT                                        { Bool_Lit($1) }
+  | LOGIC_TRUE                                      { BoolLit(true) }
```
Comment (vlad-scherbich) [inline src/parser.mly:215]: Are #215, 216 supposed to be `Bool_Lit`?

File: src/parser.mly [hunk 5]
```diff
@@ -1,27 +1,245 @@
...
+  | LOGIC_FALSE                                     { BoolLit(false) }
+  | expr ARITH_PLUS   expr                          { bin_op($1, Add, $3) }
+  | expr ARITH_MINUS  expr                          { bin_op($1, Sub, $3) }
+  | expr ARITH_TIMES  expr                          { bin_op($1, Mult, $3) }
+  | expr ARITH_DIVIDE expr                          { bin_op($1, Div, $3) }
+  | expr ARITH_MOD    expr                          { bin_op($1, Mod, $3) }
+  | expr LOGIC_EQ     expr                          { bin_op($1, Equal, $3) }
+  | expr LOGIC_NEQ    expr                          { bin_op($1, Neq, $3) }
+  | expr LCARET       expr                          { bin_op($1, Less, $3) }
```
Comment (vlad-scherbich) [inline src/parser.mly]: This is `LOGIC_LT` in the scanner.

File: src/parser.mly [hunk 6]
```diff
@@ -1,27 +1,245 @@
...
+  | expr LOGIC_LEQ    expr                          { bin_op($1, Leq, $3) }
+  | expr RCARET       expr                          { bin_op($1, Greater,$3) }
```
Comment (vlad-scherbich) [inline src/parser.mly]: This is `LOGIC_GT` in the scanner.

**Comment by Howon** [comment]
squash your commits

**Comment by Howon** [comment]
remove ast from your PR

File: src/parser.mly [hunk 7]
```diff
@@ -1,27 +1,245 @@
...
-    | LITERAL           { Lit($1) }
+  ID                                                { Id($1) }
+  | LITERAL                                         { Lit($1) }
+  | DOUBLE_LIT                                      { Double_Lit($1) }
+  | CHAR_LIT                                        { Char_Lit($1) }
+  | STRING_LIT                                      { String_Lit($1) }
+  | BOOL_LIT                                        { Bool_Lit($1) }
+  | LOGIC_TRUE                                      { BoolLit(true) }
+  | LOGIC_FALSE                                     { BoolLit(false) }
+  | expr ARITH_PLUS   expr                          { bin_op($1, Add, $3) }
```
Comment (vlad-scherbich) [inline src/parser.mly:217]: Changed to `Binop` for consistency.

**Pull request merged**